### ISSUE TITLE
Cluster does not handle certificate expiration

### ISSUE THREAD

**Issue by qbast** [banzaicloud/kafka-operator#396]
**Describe the bug**
After 3 months of cluster creation all authentication fails due to expired server certificate.

**Steps to reproduce the issue:**
Setup a cluster with SSL authentication, wait 3 months.

**Expected behavior**
I would expect certificate to be seamlessly updated.

**Additional context**
Tested using kafka-operator 0.12 (upgraded from 0.11 at some point).

I found the issue after all my kafka-dependent service failed to start, with exception containing:

`Caused by: java.security.cert.CertificateExpiredException: NotAfter: Wed Jun 10 09:25:02 GMT 2020\n\tat java.base/sun.security.x509.CertificateValidity.valid(CertificateValidity.java:274)`

I connected a debugger and confirmed the check fails on certificate for `kafka-headless.infra.svc.cluster.local.` which is the server certificate.
Then I retrieved the server certificate ( `kubectl get secret kafka-server-certificate -n infra -o yaml` ) and decoded it. 
Values I got are:
```
Common Name: kafka-headless.infra.svc.cluster.local
Subject Alternative Names: *.kafka-headless.infra.svc.cluster.local, kafka-headless.infra.svc.cluster.local, *.kafka-headless.infra.svc, kafka-headless.infra.svc, *.kafka-headless.infra, kafka-headless.infra, kafka-headless
Organization: cert-manager
Valid From: May 11, 2020
Valid To: August 9, 2020
Issuer: kafka-ca.infra.cluster.local, cert-manager
Serial Number: 4c732c295eaf6f83ba7e2996036b5bcf
```
Note validity. Also serial number is different than what I seen in the debugger.

`kubectl describe certificate kafka-headless.infra.svc.cluster.local -n infra` shows:

```
Status:
  Conditions:
    Last Transition Time:  2020-03-12T09:25:02Z
    Message:               Certificate is up to date and has not expired
    Reason:                Ready
    Status:                True
    Type:                  Ready
  Not After:               2020-08-09T09:25:02Z
```

I  assumed that pods are simply using old version of the secret and need a restart. After restarting all broker pods, cruisecontrol and operator only one kafka broker pod came up.
Logs for the lone running kafka broker shows:
```
[2020-06-19 09:32:44,163] INFO [Producer clientId=CruiseControlMetricsReporter] Failed authentication with kafka-0.kafka-headless.infra.svc.cluster.local/10.144.7.211 (SSL handshake failed) (org.apache.kafka.common.network.Selector)
[2020-06-19 09:32:44,163] ERROR [Producer clientId=CruiseControlMetricsReporter] Connection to node -1 (kafka-0.kafka-headless.infra.svc.cluster.local/10.144.7.211:9092) failed authentication due to: SSL handshake failed (org.apache.kafka.clients.NetworkClient)
[2020-06-19 09:32:44,163] WARN Failed to send Cruise Control metric [PARTITION_METRIC,PARTITION_SIZE,time=1592559064081,brokerId=0,partition=error_audit-events_audit-service-0,value=0.000] (com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter)
[2020-06-19 09:32:44,164] INFO [SocketServer brokerId=0] Failed authentication with /10.144.7.211 (SSL handshake failed) (org.apache.kafka.common.network.Selector)
[2020-06-19 09:32:44,174] INFO [SocketServer brokerId=0] Failed authentication with /10.144.8.76 (SSL handshake failed) (org.apache.kafka.common.network.Selector)
[2020-06-19 09:32:44,190] INFO [SocketServer brokerId=0] Failed authentication with /10.144.8.76 (SSL handshake failed) (org.apache.kafka.common.network.Selector)
```
10.144.7.211  is the broker itself
10.144.8.76 is kafka operator

Cruise control is crashing with :
```
11811  [| adminclient-1] ERROR he.kafka.clients.NetworkClient  - [AdminClient clientId=adminclient-1] Connection to node -1 failed authentication due to: SSL handshake failed
        at java.base/java.security.cert.CertPathValidator.validate(CertPathValidator.java:309)
        at java.base/sun.security.validator.PKIXValidator.doValidate(PKIXValidator.java:345)
        ... 39 more
Caused by: java.security.cert.CertificateExpiredException: NotAfter: Wed Jun 10 09:25:02 GMT 2020
        at java.base/sun.security.x509.CertificateValidity.valid(CertificateValidity.java:274)
        at java.base/sun.security.x509.X509CertImpl.checkValidity(X509CertImpl.java:675)
        at java.base/sun.security.provider.certpath.BasicChecker.verifyValidity(BasicChecker.java:190)
        at java.base/sun.security.provider.certpath.BasicChecker.check(BasicChecker.java:144)
        at java.base/sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:125)
```

I also connected to kafka pod (via kubectl exec) and checked the certificate file in /var/run/secrets/java.io/keystores/server/tls.crt  - it was the right one with expiration time in August. 
So in short Kafka broker has right certificate available, but it is using the old one.

**Comment by baluchicken** [comment]
Can you please check if the jks file is present in the secret? If yes can you please verify if it contains the same cert as you described?

**Comment by qbast** [comment]
Looks like you are correct: JKS is there, but it contains the old certificate with expiration date in June. Is kafka-operator responsible for generating JKS and not cert-manager?

**Comment by baluchicken** [comment]
Yeah, since the jks generation is a new feature of the cert-manager which is not available in version 0.11
